src/libcore/char.rs

```diff
@@ -448,19 +459,15 @@ impl Iterator for EscapeUnicode {
                 Some('u')
             }
             EscapeUnicodeState::LeftBrace => {
-                let mut n = 0;
-                while (self.c as u32) >> (4 * (n + 1)) != 0 {
-                    n += 1;
-                }
-                self.state = EscapeUnicodeState::Value(n);
+                self.state = EscapeUnicodeState::Value;
                 Some('{')
             }
-            EscapeUnicodeState::Value(offset) => {
-                let c = from_digit(((self.c as u32) >> (offset * 4)) & 0xf, 16).unwrap();
-                if offset == 0 {
+            EscapeUnicodeState::Value => {
+                let c = from_digit(((self.c as u32) >> (self.offset * 4)) & 0xf, 16).unwrap();
```

Comment:
Sorry, I missed a pretty simple counterexample. When iterating over 0x10000 (or any similar number with trailing zeros) after the first few hex digits the iteration would (incorrectly) stop, so the offset is needed anyway.